plugins/modem-manager/fu-sahara-loader.c

```diff
@@ -0,0 +1,644 @@
+
+		if (packet->command_id == SAHARA_HELLO_ID && packet->hello.mode < SAHARA_MODE_LAST)
+			g_debug("%s %s mode: %s", prefix, cmd_name, mode_to_str[packet->hello.mode]);
+		else if (packet->command_id == SAHARA_HELLO_RESPONSE_ID &&
+				packet->hello_response.mode < SAHARA_MODE_LAST)
+			g_debug("%s %s mode: %s", prefix, cmd_name, mode_to_str[packet->hello_response.mode]);
+		else if (packet->command_id == SAHARA_END_OF_IMAGE_TX_ID)
+			g_debug("%s %s image_id = %u, status = %s (%u)", prefix, cmd_name,
+				packet->end_of_image_transfer.image_id,
+				packet->end_of_image_transfer.status == 0 ? "SUCCESS" : "FAILED",
+				packet->end_of_image_transfer.status);
+		else
+			g_debug("%s %s", prefix, cmd_name);
+	} else
+		g_debug("%s unknown command", prefix);
+}
+
+static FuSaharaCommandId
+sahara_packet_get_command_id(GByteArray *packet)
+{
+	return ((struct sahara_packet *)(packet->data))->command_id;
+}
+
+static FuSaharaCommandId
+sahara_packet_get_length(GByteArray *packet)
+{
+	return ((struct sahara_packet *)(packet->data))->length;
+}
+
+/* initialization */
+static gboolean
+fu_sahara_loader_parse_usb_desc(FuSaharaLoader *self, GByteArray *usb_dev_desc, GError **error)
+{
+	struct usb_descriptor_header *hdr;
+	struct usb_device_descriptor *dev;
+	struct usb_interface_descriptor *ifc;
+	struct usb_endpoint_descriptor *ept;
+
+	guint8 ifclass = 0;
+	guint8 ifsubclass = 0;
+	guint8 ifnum = 0;
+
+	guint8 *ptr = usb_dev_desc->data;
+	guint8 *end = usb_dev_desc->data + usb_dev_desc->len;
+
+	/* Iterate over USB desc until we find a suitable endpoint */
+	while (ptr < end) {
+		hdr = (void *)ptr;
+
+		if (hdr->bDescriptorType == USB_DT_DEVICE) {
+			dev = (void *)ptr;
+
+			/* Devices in QDL mode always have these vid:pid */
+			if (dev->idVendor != 0x05c6 || dev->idProduct != 0x9008) {
+				g_set_error(error,
+					G_IO_ERROR,
+					G_IO_ERROR_FAILED,
+					"Wrong device and/or vendor id: 0x%04x 0x%04x",
+					dev->idVendor, dev->idProduct);
+				return FALSE;
+			}
+		} else if (hdr->bDescriptorType == USB_DT_INTERFACE && hdr->bLength == USB_DT_INTERFACE_SIZE) {
+			ifc = (void *)ptr;
+
+			ifclass = ifc->bInterfaceClass;
+			ifsubclass = ifc->bInterfaceSubClass;
+			ifnum = ifc->bInterfaceNumber;
+
+			self->in_ep = -1;
+			self->out_ep = -1;
+		} else if (hdr->bDescriptorType == USB_DT_ENDPOINT && hdr->bLength == USB_DT_ENDPOINT_SIZE ) {
+			ept = (void *)ptr;
+
+			if (ifclass == 0xff && ifsubclass == 0xff) {
+				if ((ept->bmAttributes & USB_ENDPOINT_XFERTYPE_MASK) == USB_ENDPOINT_XFER_BULK) {
+					if (ept->bEndpointAddress & USB_ENDPOINT_DIR_MASK) {
+						self->in_ep = ept->bEndpointAddress;
+						self->in_maxpktsize = ept->wMaxPacketSize;
+					} else {
+						self->out_ep = ept->bEndpointAddress;
+						self->out_maxpktsize = ept->wMaxPacketSize;
+					}
+
+					if (self->in_ep != -1 && self->out_ep != -1) {
+						self->ifnum = ifnum;
+						return TRUE;
+					}
+				}
+			}
+		}
```

Comment:
I think you want to use g_usb_interface_get_endpoints() and g_usb_endpoint_get_XXX() for all of this.